### PR TITLE
[fix] Better support for rank_zero_only setting for SLURM and torchelastic

### LINKED ISSUE #6797: LOCAL_RANK not being set in slurm
## 🐛 Bug

A lot of the PTL tooling around multiprocess depends on a specific environment variable: LOCAL_RANK being set correctly, it seems that when running in slurm this isnt set causing it to return the default of 0 for all processes which makes every process do things that should only be done on rank 0, like log stuff. 

Also I'm a little unclear about the name of that variable, if I have multiple nodes, only the *global* rank 0 not the local rank should be logging and saving checkpoints etc.

### To Reproduce

Run in slurm (cant really do it w/ colab), a good way to easily see it is to use the Wandb logger, you'll see that each process makes a new run on the Wandb UI which means that @rank_zero_experiment didnt work properly, and you can confirm this by printing LOCAL_RANK which is defaulted to 0 if unset, it will always give back 0.

### Expected behavior

LOCAL_RANK is set correctly or the rest of the tooling is aware of the global rank of the process

### Environment

Will update if it's really necessary

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #6797

This is a mitigation for the issue. The environment variable handling is currently split in a few different places:
- in this file
- in the accelerator connector
- in the cluster environment

This needs to be consolidated to better support more custom cluster environments in the future. For now this is a quickfix for SLURM users until we can come up with a better design.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -44,8 +44,18 @@ def wrapped_fn(*args, **kwargs):
     return wrapped_fn
 
 
+# TODO: this should be part of the cluster environment
+def _get_local_rank() -> int:
+    local_rank_keys = ('LOCAL_RANK', 'SLURM_LOCALID')
```
Comment (Borda) [inline pytorch_lightning/utilities/distributed.py:49]: an nasty way :D
`os.environ.get('LOCAL_RANK', os.environ.get('SLURM_LOCALID', 0))`

Comment (ananthsub) [reply pytorch_lightning/utilities/distributed.py:49]: in case we have another cluster to support and we haven't fixed the design issues here, the tuple + loop will be easier to extend. hopefully it doesn't come to that though and we can remove this altogether.

Comment (Queuecumber) [reply pytorch_lightning/utilities/distributed.py:49]: Should be SLURM_PROCID to get the global rank

File: tests/utilities/distributed.py [hunk 2]
```diff
@@ -42,3 +43,48 @@ def call_training_script(module_file, cli_args, method, tmpdir, timeout=60):
...
+
+    from pytorch_lightning.utilities.distributed import _get_rank, rank_zero_only
+    rank_zero_only.rank = _get_rank()
+
+    @rank_zero_only
+    def foo():
+        return 1
+
+    x = foo()
+    assert x is None
```
Comment (awaelchli) [inline tests/utilities/distributed.py:90]:
```suggestion
@pytest.mark.parametrize("rank_key,rank", [
    ("RANK", "1"),
    ("SLURM_PROCID", "2"),
    ("LOCAL_RANK", "3"),
])
def test_rank_zero_none_set(rank_key, rank):
    """ Test that function is not called when rank environment variables are not global zero. """

    with mock.patch.dict(os.environ, {rank_key: rank}):
        from pytorch_lightning.utilities.distributed import _get_rank, rank_zero_only
        rank_zero_only.rank = _get_rank()

        @rank_zero_only
        def foo():
            return 1

        x = foo()
        assert x is None

```
to test all possible cases?

**Review by awaelchli — COMMENTED** [review]
note, the tests won't run because the file `distributed.py` is not prefixed with `test_`

File: tests/utilities/test_distributed.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    ("RANK", "1"),
+    ("SLURM_PROCID", "2"),
+    ("LOCAL_RANK", "3"),
+])
+def test_rank_zero_none_set(rank_key, rank):
+    """ Test that function is not called when rank environment variables are not global zero. """
+
+    with mock.patch.dict(os.environ, {rank_key: rank}):
+        from pytorch_lightning.utilities.distributed import _get_rank, rank_zero_only
+        rank_zero_only.rank = _get_rank()
```
Comment (awaelchli) [inline tests/utilities/test_distributed.py:60]: seems like here setting this global variable influences other tests. 
If we run `py.test tests/utilities`, some tests fail because of this.

Comment (awaelchli) [reply tests/utilities/test_distributed.py:60]: I propose this fixture as a sanity check that the rank global variable does not leak into subsequent tests:  #6839
It simply runs before and after each test.
Tested on your branch, fixes the issue with leaking global variable.

File: tests/utilities/test_distributed.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    from pytorch_lightning.utilities.distributed import _get_rank, rank_zero_only
+    rank_zero_only.rank = _get_rank()
+
+    @rank_zero_only
+    def foo():
+        # The return type is optional because on non-zero ranks it will not be called
+        return 1
+
+    x = foo()
+    assert x == 1
```
Comment (carmocca) [inline tests/utilities/test_distributed.py:47]: why separate these 2 from the bottom one?

File: pytorch_lightning/utilities/distributed.py [hunk 5]
```diff
@@ -44,8 +44,18 @@ def wrapped_fn(*args, **kwargs):
     return wrapped_fn
 
 
+# TODO: this should be part of the cluster environment
+def _get_rank() -> int:
```
Comment (tchaton) [inline pytorch_lightning/utilities/distributed.py:48]: Can you move this directly to SLURM cluster environment ?

What if `RANK`, `SLURM_PROCID` or `LOCAL_RANK` are different ? Should we take the latest or did you order `rank_keys` based on priority ?

Best,
T.C

Comment (ananthsub) [reply pytorch_lightning/utilities/distributed.py:48]: RANK = torchelastic
SLURM_PROCID = slurm
LOCAL_RANK = parity with existing profiler

I set local rank last as RANK and SLURM_PROCID correspond to global rank already. The linked issue has more discussion, but I think we should make global rank and world size properties of the cluster environment. So the cluster environment becomes the source of truth. Cluster environment => training type plugin => accelerator => trainer. 

If the cluster environment is marked as creating children, then we can leave the initialization of these fields for later, but both torchelastic and slurm have this data already available in the environment variables, and we should expose that as soon as possible (on Trainer init) for users to read this state.

File: tests/utilities/test_distributed.py (hunk 4 above)
Comment (awaelchli) [reply tests/utilities/test_distributed.py:47]: we could parametrize the dict and then within the test do the patch with a context manager

```python
with mock.patch.dict(os.environ, env_dict):
    ...
```

**Review by awaelchli — APPROVED** [review]

File: tests/utilities/test_distributed.py (hunk 4 above)
Comment (ananthsub) [reply tests/utilities/test_distributed.py:47]: n00b question: how would we distinguish what the return should be with the parameterized dict? wouldn't we be re-implementing what's inside of `_get_rank()` ?

Comment (awaelchli) [reply tests/utilities/test_distributed.py:47]: it looks like the two tests are identical and only differ in the env variable. 
so at the moment it looks like there is no need to distinguish the output.

```python
@pytest.parametrize("env_vars", [{"RANK": "0"}, {"SLURM_PROCID": "0"}])
def test_rank_zero_known_cluster_envs():
    """ Test that torchelastic environment variables are properly checked for rank_zero_only. """

    with mock.patch.dict(os.environ, env_vars):
        from pytorch_lightning.utilities.distributed import _get_rank, rank_zero_only
        rank_zero_only.rank = _get_rank()

        @rank_zero_only
        def foo():
            # The return type is optional because on non-zero ranks it will not be called
            return 1

        x = foo()
        assert x == 1
```

Comment (ananthsub) [reply tests/utilities/test_distributed.py:47]: oh whoops, thought you meant merging with the test below for `test_rank_zero_none_set`. yes we can merge these 2 tests - I'll do that now

**Review by carmocca — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
